Simplify and update project template

kayibal opened the pull request.

kayibal committed:
simplify project template

michcio1234 commented on {{ cookiecutter.repo_name }}/deploy/docker-compose.yml:
```diff
@@ -1,68 +1,27 @@
-          dockerfile: dev.dockerfile
-
     luigid:
-        image: drtools/{{cookiecutter.repo_name}}-worker:dev
```
Should we rely on latest image, or stick to one version and maybe update it manually periodically? Not suggesting, just asking.

kayibal replied:
Hm I think it's good practice to always use the latest version

michcio1234 commented on {{ cookiecutter.repo_name }}/deploy/docker-compose.yml:
```diff
@@ -1,68 +1,27 @@
 
 
     dask-worker:
-      image: drtools/{{cookiecutter.repo_name}}-worker:dev
```
But if dask-worker doesn't have access to project code, how can it perform computations?

kayibal replied:
I think there are very little situations where the dask-worker needs to have project specific code. Project specific code should be transferred to dask with cloudpickle :)

michcio1234 replied:
I remember having some issues with it... Probably when dask worker had access to another branch of the project then the main worker. Not sure about it though.

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/structure.py:
```diff
@@ -0,0 +1,33 @@
+    EVAL_DIR = path.join(settings.ROOT, 'models/')
+    PATH_TEMPLATES.update({
+        'EVAL_DIR':EVAL_DIR,
+    })
```
Only one path in update?

kayibal replied:
Yes because I'd say paths are very project specific no?

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/structure.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def refresh():
+__factory()
```
Bug here?

kayibal replied:
Yep

michcio1234 reviewed (changes requested):
Could you please summarize these changes somehow? You removed half of the template. :P

michcio1234 commented on {{ cookiecutter.repo_name }}/setup.py:
```diff
@@ -3,14 +3,15 @@
 
 packages = find_packages()
 
+with open('deploz/requirements.txt') as fp:
```
Typo

kayibal replied:
Omg thx

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/structure.py:
```diff
@@ -0,0 +1,33 @@
+from os import path
+from drtools.conf import settings
+
+PATH_TEMPLATES = {}
```
I'd like to have constant paths separated from templates.

kayibal replied:
I was supposing to do this by naming convention. We could also add a second dictionary only called PATHS. What do you think @vitords ?

vitords replied:
I agree with @michcio1234. Seperate dictionaries for templates and constant paths will look cleaner, so +1 for the PATHS dictionary.

kayibal committed:
add PATH dictionary for constant paths

kayibal commented:
This basically removes the docker double dutch deploy routine and instead suggests to use our new automated builds for drtools which already includes dask and sparsity.

The goal is to move from a hacky docker-compose deployment (which is ment to run services not tasks) to the new deployment form where docker/kubernetes jobs/task are ran/handled via luigi. Meaning luigi actually manages docker containers which run click commands to execute certain functions. This will ease the luigi run function which just runs the docker container with a certain command.

Furthermore the project template is expected to become more general and thus fit more projects such as recommenders and scorers, which often include APIs too.

Finally we incorporate the structure module which gives an overview of paths or path templates used in the project and allows us to move from local or distributed storage by changing the ROOT key in the settings file.

kayibal commented:
@vitords 
@michcio1234

michcio1234 commented:
How do these changes affect drtools-deploy? Is the provisioning playbook still up-to-date?

kayibal commented:
No the only requirement to run this on a instance is to have the following packages installed:
- docker
- luigi
- docker-compose

michcio1234 commented:
Did we reach a conclusion on what to do with setup.py? I think a simple open('requirements.txt') will not work if we want to specify github links to our private repos in requirements file. You mentioned something about removing setup.py at all. Should we discuss it?

kayibal commented:
Yes, probably we should remove setup.py and add a dockerfile that installs the package with PYTHONPATH...

kayibal committed:
add some cli stubs and simplify cookiecutter.json

kayibal committed:
update readme

kayibal committed:
create a running example task

kayibal committed:
Unify files generated by project under root_dir.

This makes things much easier as all files will live in a single directory by default. Easing things like volume ma

kayibal committed:
Update readme

kayibal committed:
Automate versioneered git project setup

kayibal committed:
Encourage describing features with some metadata

kayibal committed:
Move external-data into data module and mark as example

kayibal committed:
Remove deploy directory

michcio1234 commented on {{ cookiecutter.repo_name }}/docker-compose.yml:
```diff
@@ -0,0 +1,41 @@
+      image: drtools/drtools:latest
+      volumes:
+        - "../data:/home/drtools/{{ cookiecutter.repo_name }}/data"
+        - "../models:/home/drtools/{{ cookiecutter.repo_name }}/models"
```
Why did you choose to map these 2 directories?

kayibal replied:
I've changed this now to map root_dir into the container. This will give us persistency and worker needs it to read and write data.

kayibal committed:
Change dask-worker mapping to use root_dir

kayibal committed:
Use relative path to root_dir

michcio1234 committed:
Add docker-compose.override, add controller service, fix some bugs (#8)

* Add docker-compose.override, add controller service, fix some bugs

* Assume that drtools dir is named drtools_base; set settings_module in relevant containers

kayibal committed:
Add external package configuration

kayibal committed:
Setup logging for package

kayibal committed:
Add optional smtp relay to send notifications

kayibal committed:
Update Readmes

michcio1234 commented on {{ cookiecutter.repo_name }}/Dockerfile:
```diff
@@ -13,4 +14,9 @@ RUN pip install -r /home/drtools/requirements.txt && \
 
+# Copy configurations into place if needed
+COPY conf/dask-config.yml /home/drtools/.dask/config.yml
+COPY conf/logging.cfg /etc/luigi/luigi-logging.cfg
```
I think this file is named `luigi-logging.cfg`, not `logging.cfg`.

michcio1234 commented on {{ cookiecutter.repo_name }}/Dockerfile:
```diff
@@ -13,4 +14,9 @@ RUN pip install -r /home/drtools/requirements.txt && \
+# Copy configurations into place if needed
+COPY conf/dask-config.yml /home/drtools/.dask/config.yml
+COPY conf/logging.cfg /etc/luigi/luigi-logging.cfg
+COPY conf/luigi.cfg /etc/luigi/client.cfg
```
I am always confused about the name and correct location of this file. But this seems to be a good solution: name it `luigi.cfg` in the template, but `client.cfg` in the image. Well, it's also somehow confusing, but I can't see a better option. :+1:

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/__init__.py:
```diff
@@ -1,19 +1,18 @@
+    try:
+        with open(log_conf_file) as fp:
+            config = yaml.safe_load(fp)
+            print(config)
```
Please add some "header" line informing what is printed and from which file.

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/__init__.py:
```diff
@@ -1,19 +1,18 @@
+        logging.config.dictConfig(config)
+    except FileNotFoundError:
+        logging.basicConfig(level=default_level)
+        logging.getLogger(__name__).warning('Error configuring logging.')
```
Please add a path to faulty file to warning message.

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/logging.yml:
```diff
@@ -0,0 +1,24 @@
+    handlers: [console]
+    qualname: {{cookiecutter.repo_name}}
+    propagate: false
+handlers:
```
This is so cool that I want more. :D
Could you please add another handler that writes to file? It shouldn't be added to any logger by default, but it would be super easy to add it. This is something I always wanted to have but never had time to do it: to have log messages displayed on console AND saved to file. Please make my dream come true. :pray:

kayibal replied:
@michcio1234  Glad you like it :)
I've been experimenting shortly with this even had it in there. But I decided to not use it as you are left with two options:

- map some shared logdir into container -> as it is the same for all containers they all write to the same file or even overwrite it?
- don't map logdir and retrieve logs from container somehow -> Too complicated... I prefer instead to use docker logs.

So I decided that what [docker logs](https://stackoverflow.com/questions/33017329/where-is-a-log-file-with-logs-from-a-container) gives us is enough to handle log files.

During this experimentation I have found some issues. Most of them I could resolve I'm still left with the issue that the command `docker-compose run controller luigi ..` currently logs everything double, but container logs show only a single log line.

michcio1234 reviewed (changes requested):
Cool! I've found one bug and some minor things to add (if you could be so kind ;) )

kayibal committed:
Fix git add remote origin

kayibal committed:
Remove print from logging setup

kayibal committed:
Add default scheduler host

kayibal committed:
Don't assume there is a dev image, unify settings.

kayibal committed:
Support running projects in parallel

kayibal committed:
Add Debug task
This task can be used to get a shell into the same environment as seen
by the task. This is useful to debug missing data or similar errors.

kayibal committed:
Normalize modules paths and root path in controller

kayibal committed:
Install docker for controller in image

kayibal committed:
Avoid nested log formatting with logs collected from containers

kayibal committed:
Fix missing handler in logging.yml

kayibal committed:
Add build section to services

kayibal committed:
Set default entrypoint to luigi --module project_name.task

kayibal committed:
Update doc

kayibal committed:
Implement FileStructure usage

kayibal committed:
Update documentation

michcio1234 commented on {{ cookiecutter.repo_name }}/README.md:
```diff
@@ -62,11 +62,29 @@
-1. Build project container: `docker build -t {{cookiecutter.repo_name}}:latest .`
-1. Start up local services: `cd {{cookiecutter.repo_name}}; docker-compose up -d`
-1. Run the example task: `luigi --module {{cookiecutter.repo_name}}.task Example`
+1. Set MODULES_PATH this is the directory that contains the current project as well as dependencies like sparsity and drtools `MODULES_PATH=..`
```
Please point out that it must be an absolute path. The example you gave suggests that it may be relative.

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings/dev.py:
```diff
@@ -9,19 +16,27 @@
+        str(_modules_path.joinpath('{{ cookiecutter.repo_name }}')):
+            {'bind': '/home/drtools/{{ cookiecutter.repo_name }}',
+             'mode': 'ro'},
+    })
```
Could you please make these 3 bindings consistent? I mean, either use `.update` one time or `__setitem__` 3 times. And either create 3 helper variables (`_sparsity`, `_drtools` and e.g. `_project`) or don't create any.

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/structure.py:
```diff
@@ -7,30 +7,17 @@
+            'CS_OUT':               ('PROCESSED', 'clickstream.pickle'),
+
+        },
+    )
```
Please use one space everywhere and add a newline at the end of file.

michcio1234 reviewed (commented):
I've just got 3 styling issues, otherwise it's supercool!

michcio1234 commented:
@vitords are you going to review it?

kayibal committed:
Cosmetics

kayibal commented:
@michcio1234 Update

michcio1234 commented on {{ cookiecutter.repo_name }}/README.md:
```diff
@@ -62,7 +62,7 @@
 ## Get Started
 
-1. Set MODULES_PATH this is the directory that contains the current project as well as dependencies like sparsity and drtools `MODULES_PATH=..`
+1. Set MODULES_PATH this is the absolute path to the directory that contains the current project as well as dependencies like sparsity and drtools `MODULES_PATH=..`
```
Hah, your example still suggests relative path but whatever... :P

kayibal replied:
Ah that's what you meant

kayibal replied:
Fixed

michcio1234 commented on {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/structure.py:
```diff
@@ -10,14 +10,13 @@
+        'PROCESSED':        ('DATA', 'processed', ),
+        'CS_OUT':               ('PROCESSED', 'clickstream.pickle'),
+    },
+)
```
And what about spaces? Doesn't pycharm underline this for you? Doesn't it disturb you? I can't stand it. :P

kayibal replied:
As you've seen I don't care about what pycharm says here :P I prefer this formatting because it suggests the structure of the directories.

kayibal replied:
see this example it's styled correctly but hard to figure out where each file will end up https://github.com/datarevenue-berlin/rtl_reco/blob/master/rtl_reco/structure.py

michcio1234 replied:
Wow, I didn't even notice that these indents actually mean something. I thought it's just messed up by tabs or something. Okay, then I can turn a blind eye to it. ;)

michcio1234 reviewed (changes requested):
No! I'm going to be a douchebag here and force style. :P

kayibal committed:
Suggest absolute path
